src/main/scala/firrtl/transforms/InlineCasts.scala

```diff
@@ -0,0 +1,59 @@
+package firrtl
+package transforms
+
+import firrtl.ir._
+import firrtl.Mappers._
+
+import firrtl.Utils.{isCast, NodeMap}
+
+object InlineCastsTransform {
+
+  /** Recursively replace [[WRef]]s with new [[Expression]]s
+    *
+    * @param replace a '''mutable''' HashMap mapping [[WRef]]s to values with which the [[WRef]]
+    * will be replaced. It is '''not''' mutated in this function
+    * @param expr the Expression being transformed
+    * @return Returns expr with [[WRef]]s replaced by values found in replace
+    */
+  def onExpr(replace: NodeMap)(expr: Expression): Expression = {
+    expr.map(onExpr(replace)) match {
+      case e @ WRef(name, _,_,_) =>
+        replace.get(name)
+               .filter(isCast)
+               .getOrElse(e)
+      case e @ DoPrim(op, Seq(WRef(name, _,_,_)), _,_) if isCast(op) =>
+        replace.getOrElse(name, e)
```

Comment:
Why are you removing casts in nested expressions?

Reply:
Oh, after reading the PR text, you answered this question: " at the moment it just ignores all casts, assuming the circuit is in a valid form, and then injects the casts for the operators that need them. " I'm not certain that this transformation is always equivalent, but I'm open to being convinced otherwise.

Reply:
I think it was a benign mistake. Combined with the behavior of the VerilogEmitter, it's fine to remove the cast, but I didn't actually mean to remove the cast here (in case someone uses this transform in another context), so I"ll fix it.

Reply:
Fixed